### PR TITLE
Lookupasn

### LINKED ISSUE #8: Replace getasn by geoipdb.LookupAsn
[geoipdb.LookupAsn](https://godoc.org/github.com/turbobytes/geoipdb#Handler.LookupAsn) contains enough features to be used as a replacement of [getasn](/turbobytes/pulse/blob/96e5a081d874dfa1f4e5228d78474abd52135688/cnc.go#L124) in cnc.go.

This is a request for this replacement to be made in cnc.go code.

### PR DESCRIPTION
**Pull request by coolparadox**
@sajal please take a look.

This also removes direct libgeoip dependency of cnc.go (remains an indirect dependency through out geoipdb however).

I tested locally (dev mode), ran a DNS test and got correct results for ASN and descriptions for a couple known ips (8.8.8.8) as expected.

Fixes #8

### CODE REVIEW

File: cnc.go [hunk 1]
```diff
@@ -557,28 +537,22 @@ func agentshandler(w http.ResponseWriter, r *http.Request) {
...
 
-func ResolveASNMtr(hop *mtrparser.MtrHop, gia *geoip.GeoIP) {
+func ResolveASNMtr(hop *mtrparser.MtrHop) {
 	hop.ASN = make([]string, len(hop.IP))
 	for idx, ip := range hop.IP {
 		//TODO...
-		hop.ASN[idx] = getasnmtr(ip, gia)
+		asn, descr, err := geo.LookupAsn(ip)
+		if err != nil {
+			hop.ASN[idx] = err.Error()
```
Comment (sajal) [inline cnc.go]: This should be blank if not available, and not an error. Check the `getasnmtr` implementation that we used earlier. This ASN info is used for each row within a single result when running a mtr test. Example: https://pulse.turbobytes.com/results/582f4cddecbe401f82000299/

Comment (coolparadox) [inline cnc.go]: @sajal Thanks for pointing this. Indeed I failed in not checking output of mtr tests after these changes, my bad.
Branch _lookupasn_ is now rebased on master, and getasnmtr returns asn identifiers only as expected. This was confirmed by checking the json output of a local mtr test in my computer (ASN json fields are ok now).
Dns tests look good also (json fields ASN and ASName are fine).

**Comment by sajal** [comment]
Hi,

I just realized there is one issue with this pull request that I should have elaborated earlier.

`github.com/turbobytes/pulse/utils` which is used by minion now imports `github.com/turbobytes/geoipdb` which imports `github.com/abh/geoip` which has a dependency on libgeoip. We cannot introduce this dependency on minion. It's tricky to cross-compile when libgeoip is a dependency, and we cannot expect existing minion hosts to make it available.

I suggest moving `LookupAsn` from utils/asnlookup.go into cnc.go . CNC is the only component that does these lookups.

**Comment by coolparadox** [comment]
+1
Sure I'll move LookupAsn as suggested so libgeoip dependency is removed from minions.

**Comment by sajal** [comment]
for future reference.

    $ go build -o /tmp/minion minion.go
    $ ldd /tmp/minion
	linux-vdso.so.1 =>  (0x00007ffdac33d000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fac2e41b000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fac2e056000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fac2e639000)

This way we can check which shared objects are linked by minion, in this case it is fine since these are standard linux libs.

In future i will get rid of these by disabling cgo in minion. Then we can have automated test to catch this issue.

**Comment by sajal** [comment]
There is one more issue. In the current implementation, when ASN cant be determined, `nil` is used. But these changes send a blank string. Please use `nil` to indicate that ASN couldn't be found because thats what the UI expects.

File: cnc.go [hunk 2]
```diff
@@ -188,6 +165,13 @@ func populatedata(w *Worker, insertfirst bool) {
 	w.FirstOnline = agent.FirstOnline
 }
 
+// lookupAsn is a wrapper around geoipdb.LookupAsn
+// that returns results as pointers
+func lookupAsn(ip string) (*string, *string, error) {
+	asn, descr, err := geo.LookupAsn(ip)
+	return &asn, &descr, err
```
Comment (sajal) [inline cnc.go]: suggestion
````
if err != nil {
    return nil, nil, err
}
````

**Comment by coolparadox** [comment]
Since nil means error, I also decided to remove the error return value from lookupasn(), so usage is simplified.

For the record, the implementation leaves logs on lookup errors, example:

```
2016/11/21 10:45:11 monitoring
2016/11/21 10:45:11 127.0.0.1:39380 at your service
2016/11/21 10:45:11 warning: libgeoip lookup failed for ip '127.0.0.1'
2016/11/21 10:45:12 warning: ipinfo lookup failed for ip '127.0.0.1': ipinfo.io lookup failed for '127.0.0.1': undefined
2016/11/21 10:45:12 warning: failed to lookup ASN for 127.0.0.1: unknown ASN for ip '127.0.0.1'
```
Let me know if this is not desired. Anyway when [geoipdb#13](/turbobytes/geoipdb/issues/13) is implemented I believe these logs for common IPs such as 127.0.0.1 can be avoided.

**Comment by sajal** [comment]
For now lets leave it as is. i.e. the old behaviour.

In the old behaviour, we treat any lookup error as "Local Network" indicated by `nil`.  Later when we tackle turbobytes/geoipdb#8, we need to explicitly detect "Local Network" and treat errors differently.

**Pull request merged**